File: internal/store/store.go
package store

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/SalzDevs/agenttop/internal/event"
)

type Store struct {
	mu       sync.Mutex
	events   []event.Event
	cap      int
	nextID   int64
	logFile  *os.File
	enc      *json.Encoder

	totalCost float64
	totalIn   int
	totalOut  int
	totalReqs int

	inFlight map[int64]bool
}

func New(path string, cap int) (*Store, error) {
	s := &Store{cap: cap, inFlight: make(map[int64]bool)}
	if path != "" {
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return nil, err
		}
		f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o600)
		if err != nil {
			return nil, err
		}
		s.logFile = f
		s.enc = json.NewEncoder(f)
	}
	return s, nil
}

func (s *Store) Close() error {
	if s.logFile != nil {
		return s.logFile.Close()
	}
	return nil
}

func (s *Store) Append(e event.Event) event.Event {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.nextID++
	e.ID = s.nextID
	if e.Time.IsZero() {
		e.Time = time.Now()
	}

	s.events = append(s.events, e)
	if len(s.events) > s.cap {
		s.events = s.events[len(s.events)-s.cap:]
	}

	if e.InFlight() {
		s.inFlight[e.TraceID] = true
	} else {
		delete(s.inFlight, e.TraceID)
		s.totalReqs++
		s.totalCost += e.CostUSD
		s.totalIn += e.InputTokens
		s.totalOut += e.OutputTokens
	}

	if s.enc != nil {
		_ = s.enc.Encode(e)
	}
	return e
}

func (s *Store) Recent(n int) []event.Event {
	s.mu.Lock()
	defer s.mu.Unlock()
	if n <= 0 || n > len(s.events) {
		n = len(s.events)
	}
	out := make([]event.Event, n)
	copy(out, s.events[len(s.events)-n:])
	return out
}

func (s *Store) Stats() (cost float64, in, out, reqs, inFlight int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.totalCost, s.totalIn, s.totalOut, s.totalReqs, len(s.inFlight)
}

File: src/store/store.ts
import { mkdirSync, openSync, closeSync, appendFileSync } from "node:fs";
import { dirname } from "node:path";
import { Event, isInFlight } from "../event/bus.js";

// Store keeps the recent event ring buffer and the running totals
// (cost, in/out tokens, request count, in-flight trace count).
// All public methods are safe to call from multiple async tasks
// concurrently; the in-memory state is guarded by a single lock.
export class Store {
  private events: Event[] = [];
  private cap: number;
  private nextId = 0;
  private logPath: string;

  private totalCost = 0;
  private totalIn = 0;
  private totalOut = 0;
  private totalReqs = 0;
  private inFlight: Set<number> = new Set();

  private lock: Promise<void> = Promise.resolve();

  constructor(logPath: string, cap: number) {
    this.cap = cap;
    this.logPath = logPath;
    if (logPath) {
      mkdirSync(dirname(logPath), { recursive: true });
      // Touch the file so the directory is created even if we never write.
      const fd = openSync(logPath, "a");
      closeSync(fd);
    }
  }

  // withLock serializes all state-mutating operations. JavaScript is
  // single-threaded, but we use a chain of promises to make multi-step
  // read-modify-write sequences atomic (e.g. emit + log + emit end).
  private async withLock<T>(fn: () => T | Promise<T>): Promise<T> {
    const prev = this.lock;
    let resolve!: () => void;
    this.lock = new Promise<void>((r) => (resolve = r));
    try {
      await prev;
      return await fn();
    } finally {
      resolve();
    }
  }

  async append(e: Event): Promise<Event> {
    return this.withLock(() => {
      this.nextId++;
      e.id = this.nextId;
      if (!e.time || isNaN(e.time.getTime())) {
        e.time = new Date();
      }

      this.events.push(e);
      if (this.events.length > this.cap) {
        this.events = this.events.slice(this.events.length - this.cap);
      }

      if (isInFlight(e)) {
        this.inFlight.add(e.traceId);
      } else {
        this.inFlight.delete(e.traceId);
        this.totalReqs++;
        this.totalCost += e.costUSD;
        this.totalIn += e.inputTokens;
        this.totalOut += e.outputTokens;
      }

      if (this.logPath) {
        try {
          appendFileSync(this.logPath, JSON.stringify(e) + "\n");
        } catch {
          // Logging must never break the proxy.
        }
      }
      return e;
    });
  }

  async recent(n: number): Promise<Event[]> {
    return this.withLock(() => {
      const k = Math.min(Math.max(n, 0), this.events.length);
      return this.events.slice(this.events.length - k);
    });
  }

  async stats(): Promise<{
    cost: number;
    in: number;
    out: number;
    reqs: number;
    inFlight: number;
  }> {
    return this.withLock(() => ({
      cost: this.totalCost,
      in: this.totalIn,
      out: this.totalOut,
      reqs: this.totalReqs,
      inFlight: this.inFlight.size,
    }));
  }
}
